MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java

```diff
@@ -1619,13 +1623,33 @@ private void updateLocation(@NonNull LocationUpdate locationUpdate, boolean from
     }
     CameraPosition currentCameraPosition = mapboxMap.getCameraPosition();
     boolean isGpsNorth = getCameraMode() == CameraMode.TRACKING_GPS_NORTH;
-    locationAnimatorCoordinator.feedNewLocation(
-      getTargetLocationWithIntermediates(locationUpdate.getLocation(), locationUpdate.getIntermediatePoints()),
-      currentCameraPosition,
-      isGpsNorth,
-      fromLastLocation ? Long.valueOf(0L) : locationUpdate.getAnimationDuration());
-    updateAccuracyRadius(locationUpdate.getLocation(), false);
-    lastLocation = locationUpdate.getLocation();
+
+
+    //Performance improvement: Check if new location requires a redraw to be performed,
+    //taking the pixel tolerance into account.
+    //By default, the tolerance is 1 pixel, meaning that the new location must impact at least
+    //a different pixel on the screen to justify a redraw.
+    boolean updateNeeded = (lastLocation == null);
+    if (updateNeeded == false) {
+      PointF lastPoint = this.mapboxMap.getProjection().toScreenLocation(
+              new LatLng(lastLocation.getLatitude(), lastLocation.getLongitude()));
+      PointF updatedPoint = this.mapboxMap.getProjection().toScreenLocation(
+              new LatLng(locationUpdate.getLocation().getLatitude(), locationUpdate.getLocation().getLongitude()));
+
+      updateNeeded = ((Math.abs(updatedPoint.x - lastPoint.x) >= options.locationTolerance())
+                      || (Math.abs(updatedPoint.y - lastPoint.y) >= options.locationTolerance()) );
```

Comment:
I can see a potential issue where:
1. we have location updates and draw the puck
2. we zoom out extremely far
3. we continue to get new location updates for some time but since the pixel distance between them is so small, we never update the puck
4. we stop location updates (or simply lose the GPS signal)
5. we zoom back in to the puck and it's drawn way off the actual last location because we kept ignoring the updates

I haven't tested this and don't know whether this is an actual risk given the range of values we're dealing with but wanted to flag regardless unless you think this could become a problem. We might want to redo the calculation every time the zoom level changes but that would undo the introduced performance improvements.

Reply:
I understand, but i don't think this append.
Today, if you move, zoom in and or out, pan.... the user position is redrawn regardless of any gps position update.
So i don't think this problem exists.

Reply:
If we didn't push the location update to the renderer in the first place, the puck would be drawn at the last valid, outdated location. This is my concern, especially if developers start to increase the tolerance.

Reply:
I understand, but the same will append if you show the map at a first valid location, then turn off your gps, then move around physically and zoom in.
In that situation due to a stop of location update, you'll see an outdated location.
With the current optimization, close the same will happend.
A tiny difference could be observed for a few seconds (location updates not taken into account), as a loss of position  would disable the feature (grey dot)
I don't think it should be a big concern, but if you think so i'll investigate an improved solution.
For the time beeing my biggest concern is the fact that the map permanetly redraws itslef as soon as the location is activated. (battery drain !)

Reply:
@LukasPaczos I updated the code to resolve somme issues. 
If necessay i can remove the ability to set tolerances.
Please, can you tell me what would be right for you in order to accept the PR ?